pandas/core/case_when.py

```diff
@@ -0,0 +1,167 @@
+        that evaluate to a 1-D boolean array/series. See examples below.
+    default : Any
+        The default value to be used if all conditions evaluate False. This value
+        will be used to create the `Series` on which `Series.mask` will be called.
+        If this value is not already an array like (i.e. it is not of type `Series`,
+        `np.array` or `list`) it will be repeated `obj.shape[0]` times in order to
+        create an array like object from it and then apply the `Series.mask`.
+
+    Returns
+    -------
+    Series
+        Series with the corresponding values based on the conditions, their values
+        and the default value.
+
+
+    Examples
+    --------
+    >>> df = pd.DataFrame(dict(a=[1, 2, 3], b=[4, 5, 6]))
+    >>> df
+       a  b
+    0  1  4
+    1  2  5
+    2  3  6
+
+    >>> pd.case_when(
+    ...     df,
+    ...     lambda x: x.a == 1,
+    ...     'first',
+    ...     lambda x: (x.a == 2) & (x.b == 5),
+    ...     'second',
+    ...     default='default',
+    ... )
+    0      first
+    1     second
+    2    default
+    dtype: object
+
+    >>> pd.case_when(
+    ...     df,
+    ...     lambda x: (x.a == 1) & (x.b == 4),
+    ...     df.b,
+    ...     default=0,
+    ... )
+    0    4
+    1    0
+    2    0
+    dtype: int64
+
+    >>> pd.case_when(
+    ...     df,
+    ...     lambda x: (x.a > 1) & (x.b > 1),
+    ...     -1,
+    ...     default=df.a,
+    ... )
+    0    1
+    1   -1
+    2   -1
+    Name: a, dtype: int64
+
+    >>> pd.case_when(
+    ...     df.a,
+    ...     lambda x: x == 1,
+    ...     -1,
+    ...     default=df.a,
+    ... )
+    0   -1
+    1    2
+    2    3
+    Name: a, dtype: int64
+
+    >>> pd.case_when(
+    ...     df.a,
+    ...     df.a  > 1,
+    ...     -1,
+    ...     default=df.a,
+    ... )
+    0    1
+    1   -1
+    2   -1
+    Name: a, dtype: int64
+
+    The index is not maintained. For example:
+    >>> df = pd.DataFrame(
+    ...     dict(a=[1, 2, 3], b=[4, 5, 6]),
+    ...     index=['index 1', 'index 2', 'index 3']
+    ... )
+    >>> df
+             a  b
+    index 1  1  4
+    index 2  2  5
+    index 3  3  6
+
+    >>> pd.case_when(
+    ...     df,
+    ...     lambda x: (x.a == 1) & (x.b == 4),
+    ...     df.b,
+    ...     default=0,
+    ... )
+    0    4
+    1    0
+    2    0
+    dtype: int64
+    """
+    len_args = len(args)
+
+    if len_args < 2:
+        raise ValueError("At least two arguments are required for `case_when`")
+    if len_args % 2:
+        raise ValueError(
+            "The number of conditions and values do not match. "
+            f"There are {len_args - len_args//2} conditions "
+            f"and {len_args//2} values."
+        )
+
+    # construct series on which we will apply `Series.mask`
+    if is_array_like(default):
+        series = pd.Series(default).reset_index(drop=True)
+    else:
+        series = pd.Series([default] * obj.shape[0])
```

Comment:
Instead of reseting the index, can you do `pd.Series(default, index=self.index)` and `pd.Series(default, index=self.index)` for each of these blocks respectively. Then I think you can remove the resetting of the index below.

Reply:
Why use is_array_like over is_list_like?

Reply:
> Why use is_array_like over is_list_like?

You are right, updated. [fd12a30](https://github.com/pandas-dev/pandas/pull/50343/commits/fd12a30cb2b40f5643a95bb09c25dfe24466a0da)

Reply:
> Instead of reseting the index, can you do pd.Series(default, index=self.index) . This can be done in the else clause as well. Then I think you can remove the resetting of the index below.

If we do so (see b897b8e), then the following case will have an undesired behavior as illustrated in this example: 

```
>>> df = pd.DataFrame(dict(a=[1, 2, 3], b=[4, 5, 6]), index=['index 1', 'index 2', 'index 3'])
>>> df
         a  b
index 1  1  4
index 2  2  5
index 3  3  6
>>> pd.case_when(
...     df,
...     lambda x: (x.a == 1) & (x.b == 4),
...     pd.Series([-1,-1,-1], index=['other index 1', 'other index 1', 'other index 1']),
...     default=0,
... )
index 1    NaN
index 2    0.0
index 3    0.0
dtype: float64
```

As a result, I overwrote the index and produced a warning to not shock the user with this index overriding. See [d01d6e8](https://github.com/pandas-dev/pandas/pull/50343/commits/d01d6e897e0a567f9d9c8e3f4bc5fc3aa18d6fe8). This behavior is similar to what was done in other places in Pandas. For example: 

https://github.com/ELHoussineT/pandas/blob/d01d6e897e0a567f9d9c8e3f4bc5fc3aa18d6fe8/pandas/core/frame.py#L3800-L3805

Reply:
If the logic looks good, kindly indicate so as I will proceed then to write the tests and docs. Otherwise, we can also continue to iterate on the logic.

Reply:
> This behavior is similar to what was done in other places in Pandas.

The behavior you're highlighting here is pandas aligning the index to `self`'s index. The logic in this PR entirely ignores the index. I would expect pandas to align the index, so that the NaN value in the first row is expected.

Reply:
Hmm.. I see. What do you suggest in this case?